Feature/add device information tlv to topology response

mariomaz opened the pull request:
[TASK] Add Device Information TLV to Topology Response message #700

mariomaz committed:
tlvf: fix DeviceInformation TLV

The nested type in `tlvDeviceInformation` was written as struct. This
was done when TLVF didn't support nested classes yet. However, it must
be a class since it has dynamically-sized content. Convert struct to
class and change its name accordingly.

Rename fields that do not match with those defined in the IEEE 1905
standard. Rename fields which type and purpose is not clear enough.

Change type of media_info field from eMediaType to array of uint8_t
since it is a variable length array which length media_info_length
and contents depend on the value of media_type field.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: autogenerate TLVs

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: rename topology query message handler method to match message name

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bcl: utils: add method to get ethernet interface speed

Ethernet interface speed is required to compute the MediaType field
in different TLVs. It is also required to compute the PHY rate when
collecting link metrics.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: add local interface field for wired interface

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -767,6 +770,54 @@ bool network_utils::linux_iface_is_up_and_running(const std::string &iface)
     return (false);
 }
 
+bool network_utils::linux_iface_get_speed(const std::string &iface, uint32_t &speed)
```
I think that returning a bool gives more information and can help distinguish the case in which speed is SPEED_UNKNOWN from that in which speed cannot be obtained. Despite this is not the case, caller might take a different decision on each case.

mariomaz commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -767,6 +770,54 @@ bool network_utils::linux_iface_is_up_and_running(const std::string &iface)
+        int rc;
+
+        beerocks::string_utils::copy_string(ifr.ifr_name, iface.c_str(), sizeof(ifr.ifr_name));
+        ifr.ifr_data = (char *)&settings;
```
Agree with you but looks like this is common practice here and C-style casts are spread all over the project. Maybe @arnout or @tomereli can give their opinion/version on this and decide if I should change it or stick to C-style

rmelotte replied:
If I may add my view, I don't think there is any good reason to use C-style casts anywhere in the project. The ones that are already there probably either were from the beginning, or slipped through during reviews.
There are also a lot of `reinterpret_cast` used all over the project, and almost/all of them could be `static_cast`s.

I think this is another good reason to add static analysis tools to at the very least catch C-style casts in C++ code, and if possible warn for `reinterpret_cast`s.

tomereli replied:
C style casts should be avoided. The ones that are there should be refactored to use c++ casts in the future.

mariomaz replied:
Done replacing C-style class with reinterpret_cast

mariomaz commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/tlvDeviceInformation.h:
```diff
@@ -63,9 +53,37 @@ class tlvDeviceInformation : public BaseClass
-        uint8_t* m_info_length = nullptr;
-        sInfo* m_info = nullptr;
-        size_t m_info_idx__ = 0;
+        uint8_t* m_local_interface_list_length = nullptr;
```
This code is auto-generated. 
Regarding this, I had some question to ask which maybe @tomereli could answer: the definition of the array of nested structures in this TLV is pretty much the same as in SupportedService TLV but auto-generated code to deal with the list of elements is different. Is there any flag to tune how to generate the code to manage the array defined in .yaml?

mariomaz commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/tlvDeviceInformation.h:
```diff
@@ -35,25 +37,13 @@ class tlvDeviceInformation : public BaseClass
-        uint8_t& info_length();
-        std::tuple<bool, sInfo&> info(size_t idx);
-        bool alloc_info(size_t count = 1);
+        uint8_t& local_interface_list_length();
```
This is auto-generated code

mariomaz commented on framework/tlvf/AutoGenerated/src/tlvf/ieee_1905_1/tlvDeviceInformation.cpp:
```diff
@@ -37,59 +37,83 @@ sMacAddr& tlvDeviceInformation::mac() {
+        TLVF_LOG(ERROR) << "No call to create_local_interface_list was called before add_local_interface_list";
+        return false;
+    }
+    uint8_t *src = (uint8_t *)m_local_interface_list;
```
This is auto-generated code

mariomaz commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -767,6 +770,54 @@ bool network_utils::linux_iface_is_up_and_running(const std::string &iface)
+        } else if (EOPNOTSUPP == errno) {
+            struct ethtool_cmd edata;
+
+            ifr.ifr_data = (char *)&edata;
```
Done too!

tomereli commented on framework/tlvf/yaml/tlvf/ieee_1905_1/tlvDeviceInformation.yaml:
```diff
@@ -10,16 +10,21 @@ tlvDeviceInformation:
+  media_info:
+    _type: uint8_t
+    _length: [ media_info_length ]
+  
```
Missing line break at end of file

mariomaz replied:
Empty line removed.

tomereli commented on framework/tlvf/AutoGenerated/include/tlvf/ieee_1905_1/tlvDeviceInformation.h:
```diff
@@ -63,9 +53,37 @@ class tlvDeviceInformation : public BaseClass
-        sInfo* m_info = nullptr;
-        size_t m_info_idx__ = 0;
+        uint8_t* m_local_interface_list_length = nullptr;
+        cLocalInterfaceInfo* m_local_interface_list = nullptr;
```
autogenerated code - but good comments, we need to take them into account when refactoring tlvf.py into TLVF version 2:)

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1776,6 +1779,24 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+        std::shared_ptr<ieee1905_1::cLocalInterfaceInfo> localInterfaceInfo =
+            tlvDeviceInformation->create_local_interface_list();
+
+        ieee1905_1::eMediaType media_type = ieee1905_1::eMediaType::UNKNONWN_MEDIA;
```
I think all this can go to a static function:
```
static bool fill_iface_info(ieee1905_1::tlvDeviceInformation &info, const std::string iface, const sMacAddr &iface_mac);
```

Then call it - `fill_iface_info(*info, bridge_info.iface, network_utils::mac_from_string(bridge_info.mac));`

mariomaz replied:
We still have to add a LocalInterfaceInfo field for each wireless interface. I'll refactor this code when I had the full picture. Promise.

tomereli reviewed (approved):
Commit 3cf23b94e162f152b3973bfce6ac466a69ff9fcf headline should be "bcl: utils: add method to get interface speed".
Also, the description should also explain a bit which TLV needs this information and for which interfaces we require to know the speed.

Other than that - LGTM, so approving. Any reason this still a draft?

ghost commented:
@mariomaz What's the status of this PR? Can it be merged soon?

mariomaz commented:
I have addressed requested changes, rebased onto master and changed the state to "Ready for review". 
"[TASK] Add Device Information TLV to Topology Response message #700" cannot be fully completed until standard commands can be sent with DWPAL (I've updated task description accordingly and added a TODO comment in code).
However, this PR could be already merged as is (if no further changes are required).

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/tlvDeviceInformation.yaml:
```diff
@@ -10,16 +10,20 @@ tlvDeviceInformation:
+    _type: uint8_t
+    _length_var: True
+  media_info:
+    _type: uint8_t
```
Why was the type changed from eMediaType to uint8_t? The commit message says that it was done, but not why.

mariomaz replied:
Sorry, I thought it was clear: **media_info** type has changed from eMediaType to uint8_t because it is an array of optional data which length **media_info_length** and contents depend on the value of the **media_type** field.

mariomaz replied:
I'm adding this explanation to the commit message

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/tlvDeviceInformation.yaml:
```diff
@@ -10,16 +10,20 @@ tlvDeviceInformation:
+    _length_var: True
+  media_info:
+    _type: uint8_t
+    _length: [ media_info_length ]
```
There should be a newline at the end of the file.

mariomaz replied:
I think I misunderstood Tomer's comment. I'm adding a newline

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1739,7 +1739,7 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
     case ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE: {
-        return handle_1905_discovery_query(cmdu_rx, src_mac);
+        return handle_1905_topology_query(cmdu_rx, src_mac);
```
Excellent that you did this in a separate commit! :1st_place_medal:

mariomaz replied:
Thanks

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1758,20 +1758,20 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
- * @brief Handles 1905 Discovery Query message, currently only prints to the log
- * @param cmdu_rx
+ * @brief Handles 1905 Topology Query message
+ * @param cmdu_rx Received CMDU (containing Topology Query)
+ * @param src_mac MAC address of the message sender
```
:1st_place_medal:

mariomaz replied:
Thanks

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1781,11 +1781,8 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
-    // https://github.com/prplfoundation/prplMesh/issues/300
-    //TODO: set number of local interfaces.
-    //TODO: fill info of each of the local interfaces, according to IEEE_1905 section 6.4.5
+    tlvDeviceInformation->mac() = network_utils::mac_from_string(bridge_info.mac);
```
There's something wrong here... The commit message title is "add DeviceInformation TLV", but the TLV was already there, and the only thing that happens is that the TODOs are removed.

I assume that this commit should be squashed with the last commit - but even then, the TODOs should probably not be removed, just rewritten.

mariomaz replied:
Done

arnout commented on common/beerocks/bcl/include/bcl/network/network_utils.h:
```diff
@@ -116,6 +116,17 @@ class network_utils {
+     * @brief Gets the speed of a network interface.
+     *
+     * @param[in] iface Name of the network interface.
+     * @param[out] speed On success, speed of the network interface as defined in SPEED_* macros
```
Excellent that you document the unit here.

Maybe add that it is in Mbps - otherwise you need to look at ethtool.h to discover that :-)

(nitpick)

mariomaz replied:
Done

arnout commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -767,6 +770,54 @@ bool network_utils::linux_iface_is_up_and_running(const std::string &iface)
+        beerocks::string_utils::copy_string(ifr.ifr_name, iface.c_str(), sizeof(ifr.ifr_name));
+        ifr.ifr_data = reinterpret_cast<char *>(&settings);
+
+        settings.cmd = ETHTOOL_GSET;
```
Unless I'm very mistaken, this must be ETHTOOL_GLINKSETTINGS. Now you're doing two times the same thing but with different structs...

mariomaz replied:
You were right. However changing it to ETHTOOL_GLINKSETTINGS was not enough (such change has uncovered a bug that I've had to fix too).

arnout reviewed (approved):
One bug, one important comment about the disappearing todos, but everything can be fixed up easily, so approving already.
